fedimint-core/src/query.rs

```diff
@@ -252,6 +256,277 @@ impl<R: Eq + Clone + Debug> QueryStrategy<R> for CurrentConsensus<R> {
     }
 }
 
+/// Query strategy that returns when all peers responded or a deadline passed
+pub struct AllOrDeadline<R> {
+    deadline: Instant,
+    num_peers: usize,
+    responses: BTreeMap<PeerId, R>,
+}
+
+impl<R> AllOrDeadline<R> {
+    pub fn new(num_peers: usize, deadline: Instant) -> Self {
+        Self {
+            deadline,
+            num_peers,
+            responses: BTreeMap::default(),
+        }
+    }
+}
+
+impl<R> QueryStrategy<R, BTreeMap<PeerId, R>> for AllOrDeadline<R> {
+    fn process(
+        &mut self,
+        peer_id: PeerId,
+        response: api::MemberResult<R>,
+    ) -> QueryStep<BTreeMap<PeerId, R>> {
+        assert!(!self.responses.contains_key(&peer_id));
+        let step = match response {
+            Ok(o) => {
+                self.responses.insert(peer_id, o);
+
+                if self.responses.len() == self.num_peers {
+                    return QueryStep::Success(mem::take(&mut self.responses));
+                }
+                QueryStep::Continue
+            }
+            // we rely on retries and timeouts to detect a deadline passing
+            Err(_e) => QueryStep::RetryMembers(BTreeSet::from([peer_id])),
+        };
+
+        if Instant::now() <= self.deadline {
```

Comment:
Not super familiar with this query process, but this `if` seems inverted.

```suggestion
        if Instant::now() > self.deadline {
```

Reply:
You're right. Thanks!